### ISSUE TITLE
MissingMethodException after patching any method with a call to built-in Unity method

### ISSUE THREAD

**Issue by phobos2077** [pardeike/Harmony#376]
**Describe the bug**
Patching any method containing a call to Unity built-in functions (such as GameObject.Find or Random.Range) results in a MissingMethodException on an attempt to call this built-in function from ANY other method!

**To Reproduce**
1. The original method and its signature and class
```
  private Room CreateNewLargeRoom()
  {
    Room room = new Room();
    if (this.InternalMapInfo.SpawnRoom == null)
    {
      room.xDimention = this.SpawnRoomSize;
      room.zDimention = this.SpawnRoomSize;
    }
    else
    {
      room.xDimention = Random.Range(this.MinRoomSize, this.MaxRoomSize + 1);
      room.zDimention = Random.Range(this.MinRoomSize, this.MaxRoomSize + 1);
    }
    return room;
  }
```

2. The patch code
```
    [HarmonyPatch(typeof(MapGenerator))]
    [HarmonyPatch("CreateNewLargeRoom")]
    class PatchTestTranspile
    {
        static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
        {
            foreach (var instr in instructions) yield return instr;
        }
    }
```

3. The output of the Harmony debug log
```
### Harmony id=com.testmod.patch, version=2.0.4.0, location=D:\GAMES\SteamLibrary\steamapps\common\COMPOUND\0Harmony.dll, env/clr=4.0.30319.17020, platform=Win32NT
### Started from static System.Void TestMod.MyPatcher::DoPatching(), location D:\GAMES\SteamLibrary\steamapps\common\COMPOUND\TestMod.dll
### At 2021-01-14 03.49.36
### Patch: Room MapGenerator::CreateNewLargeRoom()
### Replacement: static Room MapGenerator::MapGenerator.CreateNewLargeRoom_Patch0(MapGenerator this)
IL_0000: Local var 0: Room
IL_0000: // start original
IL_0000: newobj     System.Void Room::.ctor()
IL_0005: stloc.0
IL_0006: ldarg.0
IL_0007: ldfld      MapInfo MapGenerator::InternalMapInfo
IL_000C: ldfld      Room MapInfo::SpawnRoom
IL_0011: brtrue =>  Label0
IL_0016: ldloc.0
IL_0017: ldarg.0
IL_0018: ldfld      System.Int32 MapGenerator::SpawnRoomSize
IL_001D: stfld      System.Int32 Room::xDimention
IL_0022: ldloc.0
IL_0023: ldarg.0
IL_0024: ldfld      System.Int32 MapGenerator::SpawnRoomSize
IL_0029: stfld      System.Int32 Room::zDimention
IL_002E: br =>      Label1
IL_0033: Label0
IL_0033: ldloc.0
IL_0034: ldarg.0
IL_0035: ldfld      System.Int32 MapGenerator::MinRoomSize
IL_003A: ldarg.0
IL_003B: ldfld      System.Int32 MapGenerator::MaxRoomSize
IL_0040: ldc.i4.1
IL_0041: add
IL_0042: call       static System.Int32 UnityEngine.Random::Range(System.Int32 min, System.Int32 max)
IL_0047: stfld      System.Int32 Room::xDimention
IL_004C: ldloc.0
IL_004D: ldarg.0
IL_004E: ldfld      System.Int32 MapGenerator::MinRoomSize
IL_0053: ldarg.0
IL_0054: ldfld      System.Int32 MapGenerator::MaxRoomSize
IL_0059: ldc.i4.1
IL_005A: add
IL_005B: call       static System.Int32 UnityEngine.Random::Range(System.Int32 min, System.Int32 max)
IL_0060: stfld      System.Int32 Room::zDimention
IL_0065: Label1
IL_0065: ldloc.0
IL_0066: // end original
IL_0066: ret
DONE
```

4. The stacktrace or other errors
```
MissingMethodException: Attempted to access a missing method.
  at (wrapper managed-to-native) UnityEngine.Random.RandomRangeInt(int,int)
  at UnityEngine.Random.Range (System.Int32 min, System.Int32 max) [0x00001] in <d406ff682e424fcfa5290bc9a41921af>:0 
  at FootStepEmitter.Update () [0x0006d] in <eeeb1103330b440b81839932afc0c0e1>:0 
```

```
MissingMethodException: Attempted to access a missing method.
  at (wrapper managed-to-native) UnityEngine.Random.RandomRangeInt(int,int)
  at UnityEngine.Random.Range (System.Int32 min, System.Int32 max) [0x00001] in <d406ff682e424fcfa5290bc9a41921af>:0 
  at RandomClipPlayer.PlayRandomClip () [0x00018] in <eeeb1103330b440b81839932afc0c0e1>:0 
  at DropletSounds.OnParticleCollision (UnityEngine.GameObject other) [0x00019] in <eeeb1103330b440b81839932afc0c0e1>:0 
```
As you see those are 2 different **unpatched** methods that now produce an exception.

**Expected behavior**
Built-in Unity method calls should not produce the aforementioned exception.

**Screenshots / Code**
-

**Runtime environment (please complete the following information):**
 - OS: Windows 10, 64bit
 - .NET version: 4.8
 - Harmony version 2.0.4
- Name of game or host application: COMPOUND (Unity 2019.3.10)

**Additional context**
Have anyone tested this lib with more recent versions of Unity? It seems to me there might be some incompatibility issue.

**Comment by phobos2077** [comment]
Original discussion: https://github.com/pardeike/Harmony/issues/374#issuecomment-759826964

**Comment by phobos2077** [comment]
The same patcher library works flawlessly (prefix/postfix works) and doesn't break the game if I find a method that doesn't directly call any Unity functions.

**Comment by podostro** [comment]
@phobos2077: I found a workaround for my similar issue... maybe you could do the same...

You're probably executing Harmony's patching from your assembly's main method like I was doing.
Postponing this to execute after at least the first game scene was loaded, seems to solve it.

I came upon this after realizing the cause I wasn't able to create a GameObject from code in my assembly was this very issue; and fixed it by postponing my GameObject creation to after the SceneManager.sceneLoaded event is called.

e.g:

```
    public static class Patcher
    {
        private static readonly string HarmonyId = $"Mod v{ModConstants.ModVersions.ModVersion_Current}";
        private static bool patched = false;

        public static void Main()
        {
            try
            {
                if (!ConfigProvider.Config.EnableMod)
                {
                    ModLogger.Log("Mod is disabled");

                    return;
                }

                //DoPatch();  <-- Don't patch here, postpone it till Unity engine calls SceneManager.sceneLoaded event()
                UnityEngine.SceneManagement.SceneManager.sceneLoaded += SceneLoaded;
            }
            catch (Exception ex)
            {
                ModLogger.Log(ex.ToString());
            }
        }

        private static void DoPatch()
        {
            var harmony = new Harmony(HarmonyId);
            harmony.PatchAll();

            patched = true;

            ModLogger.Log("Patched");
        }

        private static void SceneLoaded(UnityEngine.SceneManagement.Scene scene, UnityEngine.SceneManagement.LoadSceneMode loadSceneMode)
        {
            ModLogger.Log($"SceneLoaded. Scene: {scene.name}. loadSceneMode: {loadSceneMode}");

            if (!patched)
                DoPatch(); // <-- patch here.

        }
    }

```


@pardeike : I'm guessing Unity does dynamic linking of the "extern" methods at start up, which seem to cause the problem since the patched code at my original point in time don't know the actual final linked method.
Clarifying: 
- MonoBehaviour.Invoke() is an external method, and it's the one which calls Awake() (my original problem)
- The RandomRangeInt causing phobos problem is also an external method.

So to sum up: Harmony is not the cause of the problem XD
(Maybe add this workaround as a suggestion to the docs?)


![image](https://user-images.githubusercontent.com/2327913/104775694-8a5c2d00-5757-11eb-8239-29692ed9857f.png)
![image](https://user-images.githubusercontent.com/2327913/104775797-bb3c6200-5757-11eb-89d9-28c7d8a1a243.png)

**Comment by pardeike** [comment]
Good call! Since it's Unity and therefore a common case (although I haven't heard about it yet) it is worth mentioning in the docs. Probably here (hint: PR welcome): https://github.com/pardeike/Harmony/blob/master/Harmony/Documentation/articles/patching-edgecases.md

**Comment by phobos2077** [comment]
@podostro wow, great thinking man!! I will try it out. Sounds like a reasonable explanation. I wonder why it did work for other games (versions of Unity?). Maybe the reason of my issue is in the patcher I'm using (https://github.com/NeighTools/UnityDoorstop/wiki), which likely hooks too early resulting in broken extern methods? I still don't understand exactly how it happens. Does Unity modify these extern methods somehow during engine load? How does it explain that after patching ALL existing (non-patched) code gets broken due to "corrupted" extern methods?

**Comment by podostro** [comment]
@pardeike : PR: [379](https://github.com/pardeike/Harmony/pull/379)

@phobos2077: I'm also using Doorstop, it's main selling point is:
> Doorstop executes the assembly before Unity has the ability to modify its application domain

To understand what this means, and what I meant by Unity dynamically linking those `extern` methods, you can search for "static vs dynamic library linking", dynamic linking is a technique which allows for modular updating of the application's binaries.

I'm guessing other games (and older Unity versions) didn't use dynamic linking. 
I'm assuming the [GeneratedByOldBindingsGenerator] attribute you can see on the decompiled code was added by Unity to identity such cases.

```csharp
[MethodImpl(MethodImplOptions.InternalCall)]
[GeneratedByOldBindingsGenerator]
public static extern void DontDestroyOnLoad(Object target);
```

**Comment by phobos2077** [comment]
Preliminary test (with Transpiling one method) shows that your solution indeed works. I will test by patching more methods that use external functions and report back.

**Comment by podostro** [comment]
@pardeike : Is there a chance Harmony handles such cases when patching a method? 

I don't know IL well enough..., but I assume when reading the original IL the `call` opcode will have a different argument type than general methods calls. 
Is it possible for Harmony to see if the method to call is `external` and if it is actually linked at the moment of patching?

In case this is possible, it would be great if patching threw an exception; so we can know when such cases occurs early on.

**Comment by pardeike** [comment]
That's a very specific feature that would slow down 95% of the normal code cases. Harmony usually does not bother with assembler loading and linking problems because they are out of scope for the project.

I refer to other solutions that use Harmony on a higher level instead.

**Comment by phobos2077** [comment]
Tried patching other method. Everything seems to work. I strongly suggest updating the documentation with this caveat and suggested workaround as I think many users will stumble upon this just as I did as a completely fresh user. @pardeike
